File: python/exer05a_product_matrix_vector.py
'''
MATRIX-VECTOR MULTIPLICATION
'''

import threading



def get_scalar_product(u: list, v: list):
    s = sum(a * b for a, b in zip(u, v))
    return s



def scalar_thfunc(u: list, v: list, res: list, idx_res: int):
    scalar_prod = get_scalar_product(u, v)
    res[idx_res] = scalar_prod



def get_product(mat: list[list], vec: list) -> list:
    # Assume that size of mat and vec are both eligible
    size_row_mat = len(mat)
    # size_col_mat = len(mat[0])
    # size_vec = len(vec)

    res = [0] * size_row_mat
    lstth = []

    for i in range(size_row_mat):
        lstth.append(threading.Thread(target=scalar_thfunc, args=(mat[i], vec, res, i)))

    for th in lstth:
        th.start()

    for th in lstth:
        th.join()

    return res



A = [
    [ 1, 2, 3 ],
    [ 4, 5, 6 ],
    [ 7, 8, 9 ]
]

b = [
    3,
    -1,
    0
]

result = get_product(A, b)
print(result)

File: cpp/cpp-std/exer05a-product-matrix-vector.cpp
/*
MATRIX-VECTOR MULTIPLICATION
*/


#include <iostream>
#include <vector>
#include <thread>
#include "exer05-util.hpp"
using namespace std;



using vectord = std::vector<double>;
using matrix = std::vector<vectord>;



void getProduct(const matrix& mat, const vectord& vec, vectord& result) {
    // Assume that size of mat and vec are both eligible
    int sizeRowMat = mat.size();
    int sizeColMat = mat[0].size();
    int sizeVec = vec.size();

    result.clear();
    result.resize(sizeRowMat, 0);

    std::vector<std::thread> lstTh(sizeRowMat);

    for (int i = 0; i < sizeRowMat; ++i) {
        auto&& u = mat[i].data();
        auto&& v = vec.data();
        lstTh[i] = std::thread(&getScalarProduct, u, v, sizeVec, &result[i]);
    }

    for (auto&& th : lstTh) {
        th.join();
    }
}



int main() {
    matrix A = {
        { 1, 2, 3 },
        { 4, 5, 6 },
        { 7, 8, 9 }
    };

    vectord b = {
        3,
        -1,
        0
    };

    vectord result;
    getProduct(A, b, result);

    for (int i = 0; i < result.size(); ++i) {
        cout << result[i] << endl;
    }

    return 0;
}
